### ISSUE TITLE
Chrome/Safari rounding / alignment bug [was: problem with \overline]

### ISSUE THREAD

**Issue by slibedinsky** [mathjax/MathJax#1498]
When I write in mathjax: 
$3,\overline {4}$ it shows the number 4 below what's supposed to be like shown in the image (instead of being alined with the rest of the number):

![screenshot 2016-06-01 11 57 00](https://cloud.githubusercontent.com/assets/19683827/15704030/55b098d8-27f1-11e6-93fc-dda5607104cd.png)

**Comment by pkra** [comment]
I can't reproduce this. Do you see this on http://cdn.mathjax.org/mathjax/latest/test/sample-dynamic-2.html as well? Does it change with the browser zoom-level?

**Comment by libedinsky** [comment]
There I don't see any problems! Thank you so much!

**Comment by pkra** [comment]
Ok, that seems to indicate it is a problem specific to the demo on mathjax.org. We've had #1393 which affected the site but we added a workaround. Could you try to clear your browser cache for mathjax.org?

**Comment by libedinsky** [comment]
I tried, but i'm not sure how to errase the browser cache only for mathjax.org. I tried this: http://superuser.com/questions/278948/clear-cache-for-specific-domain-name-in-chrome/598260#598260

but couldn't find the "gear" icon

**Comment by pkra** [comment]
On Chrome, if you open the dev tools (F12), you can long- or right-click on the normal reload button for more options.

**Comment by libedinsky** [comment]
I did it, but it still appears in the wrong position in the demo.

**Comment by pkra** [comment]
Thanks for trying. Could you tell us what Chrome and OS version you are on?

**Comment by libedinsky** [comment]
OS X Yosemite version 10.10.5 and Chrome Version 50.0.2661.102 (64-bit).
Thanks again!

**Comment by dpvc** [comment]
I can reproduce it in Chrome on Mac OS.  It seems to be Chrome-specific, and it only affects the CommonHTML output, which is why the sample-dynamic-2 demo doesn't show it, because it used HTML-CSS by default; if you change to CommonHTML, it also shows the issue.

**Comment by pkra** [comment]
>  if you change to CommonHTML, it also shows the issue.

D'oh. Sorry about causing confusion. FWIW, I didn't see it on mathjax.org -- but take your word for it, @dpvc ;-)

**Comment by dpvc** [comment]
This turns out to be a Chrome bug where it doesn't handle rounding properly in some cases.  In particular, it seems that the height of certain boxes are rounded up, but the positioning of the top doesn't take that into account properly.

Consider the following HTML document.  In it, we display an "x" with a red background and next to it a span that has `display:inline-block`.  Inside that inline-block are two spans with `display:block`, the second one being an "x" on the pink background, and the first a span with a thin height and green background.  These should be stacked vertically, and the whole in-line block should be aligned on the baseline of the bottom span in the inline-block.  That is, the two x's should be aligned.

There are two copies of this setup:  one where the green block has height 2.5px and one where its height is 2.499px.  That is, the first one's height should round up while the second should round down.  This does seem to be the case, but the tops of the inline-block elements are not adjust to compensate.  So the first pink "x" is one pixel too low, as shown in the diagram below:

![chrome-bug](https://cloud.githubusercontent.com/assets/432782/15878531/a5cc2556-2ce9-11e6-8d5d-ba5a72f88b67.png)

Here, the diagram is enlarged to make the pixel difference more visible.  Note that the second pair is aligned while the first is not, and that the green block is rounded up to 3 pixels.

```
<!DOCTYPE html>
<html>
<head>
<title>Chrome alignment bug</title>
</head>
<body>

<div style="font-family:Times; font-size: 24px">

<span style="background-color:red">x</span><span id="span1" style="display:inline-block">
  <span style="display:block; height:2.5px; background-color:green"></span>
  <span style="display:block; background-color:pink">x</span>
</span>

<span style="background-color:red">x</span><span id="span2" style="display:inline-block">
  <span style="display:block; height:2.4999px; background-color:green"></span>
  <span style="display:block; background-color:pink">x</span>
</span>

</div>

<pre id="output">
</pre>

<script>
var s1 = document.getElementById("span1");
var s2 = document.getElementById("span2");
var output = document.getElementById("output");
var b1 = s1.getBoundingClientRect();
var b2 = s2.getBoundingClientRect();
output.innerHTML = [
  "span1:",
  "  top:          "+b1.top,
  "  bottom:       "+b1.bottom,
  "  height:       "+b1.height,
  "  offsetHeight: "+s1.offsetHeight,
  "",
  "span2:",
  "  top:          "+b2.top,
  "  bottom:       "+b2.bottom,
  "  height:       "+b2.height,
  "  offsetHeight: "+s2.offsetHeight
].join("\n");
</script>
</body>
</html>
```

The HTML document also prints out information about the sizes of the two inline-block elements:

```
span1:
  top:          8
  bottom:       38.5
  height:       30.5
  offsetHeight: 31

span2:
  top:          8
  bottom:       38.484375
  height:       30.484375
  offsetHeight: 30
```

Note that the tops are both 8, while the offsetHeights differ by 1.  In order to be aligned properly, the top of span1 should have been 7.5 (rounded down to 7), but it is incorrectly set to 8.

Note that issue is also dependent on the zoom level, and if you zoom in and out, you can get the first one to be aligned correctly and the second to misalign.

I don't think there is anything MathJax can do about this bug.

**Comment by dpvc** [comment]
I'm dropping this from the milestone, since I don't think there is anything we can do to fix it.

**Comment by pkra** [comment]
:+1:

**Comment by pkra** [comment]
@dpvc I was working on a bug report and couldn't fully reproduce this.

* setup: http://codepen.io/pkra/pen/mEqOGx?editors=1010
* On Chrome 51.0.2704.106 I do get a difference in offsetHeight at 100% zoom level (all others are the same)
* the rendering looks like below (which seems different from your screenshot)

![screenshot](https://cloud.githubusercontent.com/assets/426686/16765153/b545d8ce-4832-11e6-9788-f3d5eb1b23b1.png)

**Comment by pkra** [comment]
Oh, I see now that FF and WebKit look differently, in particular the bottom of the pink and red box fit.

**Comment by pkra** [comment]
Ok, one more. In WebKit, the bottom of the pink box "jumps around" when zooming in and out (sometimes lower than the red box). FF seems pretty stable.

**Comment by dpvc** [comment]
Yes, this is a WebKit-specific bug (I thought it was Chrome-specific, but in June I was using an older version of Safari from before they changed to decimal parts of a pixel, and so it didn't show up there).  

Yes, zooming does cause the results to vary (as I mentioned at the bottom of my post).

One reason your results are different is because of the titles at the top of your page.  That puts the boxed characters at a non-integer pixel position, and that affects the results.  You could make the enclosing `<div>` have `position:absolute; top:120px;` for example, and push the `<pre>` element into the `<div>` so that it is not overlapped.  You would then want to add `font-size:70%` or perhaps smaller to the `<pre>`.  These changes will put the test case on a pixel boundary, where the results should be more consistent with my example above.

**Comment by dpvc** [comment]
You might also want to prevent the top text from wrapping, since if it does, it will be overlapped by the absolutely positioned test div.

**Comment by pkra** [comment]
Thanks, @dpvc. I've updated the demo and will submit bug reports later.

**Comment by pkra** [comment]
I never filed a bug. Will try to remedy that.

**Comment by pkra** [comment]
Bugs filed at https://bugs.webkit.org/show_bug.cgi?id=164312 and https://bugs.chromium.org/p/chromium/issues/detail?id=661468

**Comment by dpvc** [comment]
This browser bug was fixed in Chrome, but remains in WebKit.  We have done what we can to work around it, so I'm closing this issue for now.
